docs/configuration_reference.rst
Configuration reference
=======================

All available configuration options are listed below with their default values.

.. code-block:: yaml

    FOS_user:
        db_driver:              ~ # Required
        firewall_name:          ~ # Required
        user_class:             ~ # Required
        use_listener:               true
        use_flash_notifications:    true
        model_manager_name:         null  # change it to the name of your entity/document manager if you don't want to use the default one.
        use_authentication_listener: true
        from_email:             ~ # Required
        resetting:
            retry_ttl: 7200 # Value in seconds, logic will use as hours
            token_ttl: 86400
            from_email: # Use this node only if you don't want the global email address for the resetting email
        service:
            mailer:                 FOS_user.mailer.default
            email_canonicalizer:    FOS_user.util.canonicalizer.default
            username_canonicalizer: FOS_user.util.canonicalizer.default
            token_generator:        FOS_user.util.token_generator.default
            user_manager:           FOS_user.user_manager.default
        group:
            group_class:    ~ # Required when using groups
            group_manager:  FOS_user.group_manager.default


CHANGELOG.md
# 1.6.1

## Changes

## 🐛 Bug Fixes

- Fix button translation [@core23] ([#180])

# 1.6.0

## Changes

- Move submit buttons to form definition [@core23] ([#164])

## 🚀 Features

- Replace hardcoded forms with symfony forms [@core23] ([#167])

## 🐛 Bug Fixes

- Fix `PreUpdateEventArgs` import for mongodb [@core23] ([#176])
- Fix loading driver related validations [@core23] ([#173])

## 📦 Dependencies

- Add support for PHP 8 [@core23] ([#148])

# 1.5.0

## Changes

## 🐛 Bug Fixes

- Add more specific phpdoc [@core23] ([#120])
- Add more specific phpdoc [@core23] ([#141])

## 📦 Dependencies

- Add doctrine/common 3 support [@TorbenLundsgaard] ([#137])
- Drop doctrine/mongodb-odm 1 [@TorbenLundsgaard] ([#136])

# 1.4.0

## Changes

## 🚀 Features

- Adding Spanish translations [@anacona16] ([#100])
- Move configuration to PHP [@core23] ([#52])

## 🐛 Bug Fixes

- Add setPlainPassword() before update user [@core23] ([#92])
- Remove deprecated mongo safe option [@core23] ([#50])
- Catch possible null error [@core23] ([#49])

# 1.3.1

## 🐛 Bug Fixes

- Disable setting twig timezone [@core23] ([#40])

# 1.3.0

## Changes

## 🐛 Bug Fixes

- Find methods could return null [@core23] ([#38])

# 1.2.0

## Changes

## 🚀 Features

- Use bootstrap 3 layout as default form theme [@core23] ([#36])

# 1.1.0

## Changes

## 🚀 Features

- Add noop driver to allow flex recipe [@core23] ([#34])
- Add missing toString method to Group model [@core23] ([#32])
- Make service aliases public [@core23] ([#30])

## 🐛 Bug Fixes

- Prefix generics with [@phpstan] [@core23] ([#31])

[#141]: https://github.com/Nucleos/NucleosUserBundle/pull/141
[#137]: https://github.com/Nucleos/NucleosUserBundle/pull/137
[#136]: https://github.com/Nucleos/NucleosUserBundle/pull/136
[#120]: https://github.com/Nucleos/NucleosUserBundle/pull/120
[#100]: https://github.com/Nucleos/NucleosUserBundle/pull/100
[#92]: https://github.com/Nucleos/NucleosUserBundle/pull/92
[#52]: https://github.com/Nucleos/NucleosUserBundle/pull/52
[#50]: https://github.com/Nucleos/NucleosUserBundle/pull/50
[#49]: https://github.com/Nucleos/NucleosUserBundle/pull/49
[#40]: https://github.com/Nucleos/NucleosUserBundle/pull/40
[#38]: https://github.com/Nucleos/NucleosUserBundle/pull/38
[#36]: https://github.com/Nucleos/NucleosUserBundle/pull/36
[#34]: https://github.com/Nucleos/NucleosUserBundle/pull/34
[#32]: https://github.com/Nucleos/NucleosUserBundle/pull/32
[#31]: https://github.com/Nucleos/NucleosUserBundle/pull/31
[#30]: https://github.com/Nucleos/NucleosUserBundle/pull/30
[@phpstan]: https://github.com/phpstan
[@core23]: https://github.com/core23
[@anacona16]: https://github.com/anacona16
[@TorbenLundsgaard]: https://github.com/TorbenLundsgaard
[#176]: https://github.com/Nucleos/NucleosUserBundle/pull/176
[#173]: https://github.com/Nucleos/NucleosUserBundle/pull/173
[#167]: https://github.com/Nucleos/NucleosUserBundle/pull/167
[#164]: https://github.com/Nucleos/NucleosUserBundle/pull/164
[#148]: https://github.com/Nucleos/NucleosUserBundle/pull/148
[#180]: https://github.com/Nucleos/NucleosUserBundle/pull/180


src/DependencyInjection/Configuration.php
<?php

declare(strict_types=1);

/*
 * This file is part of the FOSUserBundle package.
 *
 * (c) Christian Gripp <mail@core23.de>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace FOS\UserBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\NodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

final class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('fos_user');

        $rootNode = $treeBuilder->getRootNode();

        $this->addMainSection($rootNode);
        $this->addResettingSection($rootNode);
        $this->addGroupSection($rootNode);
        $this->addServiceSection($rootNode);

        return $treeBuilder;
    }

    private function addMainSection(NodeDefinition $node): void
    {
        $supportedDrivers = ['noop', 'orm', 'mongodb', 'custom'];

        $node
            ->children()
                ->scalarNode('db_driver')
                    ->validate()
                        ->ifNotInArray($supportedDrivers)
                        ->thenInvalid('The driver %s is not supported. Please choose one of '.json_encode($supportedDrivers))
                    ->end()
                    ->cannotBeOverwritten()
                    ->defaultValue('noop')
                    ->cannotBeEmpty()
                ->end()
                ->scalarNode('user_class')->isRequired()->cannotBeEmpty()->end()
                ->scalarNode('firewall_name')->isRequired()->cannotBeEmpty()->end()
                ->scalarNode('model_manager_name')->defaultNull()->end()
                ->booleanNode('use_authentication_listener')->defaultTrue()->end()
                ->booleanNode('use_listener')->defaultTrue()->end()
                ->booleanNode('use_flash_notifications')->defaultTrue()->end()
                ->scalarNode('from_email')->isRequired()->cannotBeEmpty()->end()
            ->end()
            // Using the custom driver requires changing the manager services
            ->validate()
                ->ifTrue(static function ($v): bool {
                    return 'custom' === $v['db_driver'] && 'fos_user.user_manager.default' === $v['service']['user_manager'];
                })
                ->thenInvalid('You need to specify your own user manager service when using the "custom" driver.')
            ->end()
            ->validate()
                ->ifTrue(static function ($v): bool {
                    return 'custom' === $v['db_driver'] && isset($v['group']) && 'fos_user.group_manager.default' === $v['group']['group_manager'];
                })
                ->thenInvalid('You need to specify your own group manager service when using the "custom" driver.')
            ->end()
        ;
    }

    private function addResettingSection(NodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('resetting')
                    ->addDefaultsIfNotSet()
                    ->canBeUnset()
                    ->children()
                        ->scalarNode('retry_ttl')->defaultValue(7200)->end()
                        ->scalarNode('token_ttl')->defaultValue(86400)->end()
                        ->scalarNode('from_email')->end()
                    ->end()
                ->end()
            ->end()
        ;
    }

    private function addServiceSection(NodeDefinition $node): void
    {
        $node
            ->addDefaultsIfNotSet()
            ->children()
                ->arrayNode('service')
                    ->addDefaultsIfNotSet()
                        ->children()
                            ->scalarNode('mailer')->defaultValue('fos_user.mailer.default')->end()
                            ->scalarNode('email_canonicalizer')->defaultValue('fos_user.util.canonicalizer.default')->end()
                            ->scalarNode('token_generator')->defaultValue('fos_user.util.token_generator.default')->end()
                            ->scalarNode('username_canonicalizer')->defaultValue('fos_user.util.canonicalizer.default')->end()
                            ->scalarNode('user_manager')->defaultValue('fos_user.user_manager.default')->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }

    private function addGroupSection(NodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('group')
                    ->canBeUnset()
                    ->children()
                        ->scalarNode('group_class')->isRequired()->cannotBeEmpty()->end()
                        ->scalarNode('group_manager')->defaultValue('fos_user.group_manager.default')->end()
                    ->end()
                ->end()
            ->end()
        ;
    }
}
